src/server/journal/journal_slice.cc

```diff
@@ -116,36 +117,63 @@ error_code JournalSlice::Close() {
 }
 #endif
 
+bool JournalSlice::IsLSNInBuffer(LSN lsn) const {
+  DCHECK(ring_buffer_);
+
+  if (ring_buffer_->empty()) {
+    return false;
+  }
+  return (*ring_buffer_)[0].lsn <= lsn && lsn <= ((*ring_buffer_)[ring_buffer_->size() - 1].lsn);
+}
+
+std::string_view JournalSlice::GetEntry(LSN lsn) const {
+  DCHECK(ring_buffer_ && IsLSNInBuffer(lsn));
+  auto start = (*ring_buffer_)[0].lsn;
+  DCHECK((*ring_buffer_)[lsn - start].lsn == lsn);
+  return (*ring_buffer_)[lsn - start].data;
+}
+
 void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
-  // DCHECK(ring_buffer_);
-  if (entry.opcode != Op::NOOP) {
-    lsn_++;
-// TODO: This is preparation for AOC style journaling, currently unused.
-#if 0
-    RingItem item;
-    item.lsn = prev_lsn;
+  DCHECK(ring_buffer_);
+
+  JournalItem dummy;
+  JournalItem* item;
+  if (entry.opcode == Op::NOOP) {
+    item = &dummy;
+    item->lsn = -1;
+    item->opcode = entry.opcode;
+    item->data = "";
+  } else {
+    item = ring_buffer_->GetTail(true);
+    item->opcode = entry.opcode;
+    item->lsn = lsn_++;
 
-    item.opcode = entry.opcode;
-    item.txid = entry.txid;
-    VLOG(1) << "Writing item [" << item.lsn << "]: " << entry.ToString();
-    ring_buffer_->EmplaceOrOverride(move(item));
+    io::BufSink buf_sink{&ring_serialize_buf_};
```

Comment:
I think using StringSink is better here.
you serialize into a string and then just move into `item->data` saving a copy at line 155.

Reply:
Keeping a big backing buffer will prevent grow allocations

Reply:
I can also just serialize directly into `item->data` without moving, no?

Reply:
nvm, StringSink owns its string

Reply:
So - if I always move out of a StringBuf I always have to allocate a new string. Where as currently I'm reusing the strings from the ring buffer. I think less allocations is better than less copies in this case

Reply:
you always alllocate a new string with `BufSink` as well: `item->data = io::View(ring_serialize_buf_.InputBuffer());`

Reply:
Why? `item` is a pointer to an object on the ring buffer and I'm assigning to it a new string.

Reply:
I don't understand where any allocation happens here (except for the first 16k entries)

Reply:
I did not remember we reuse items in the ring buffer. It's worth adding a comment because it can bite us later by hiding allocated memory.

Reply:
So, I thought about this again. I agree that memory usage can be an issue. I like @dranikpg's suggestion about two buffers (one for entries, one for bytes) but it's a bit more tricky to handle.
So I suggest for now to lower the default size (which is currently 64k, I didn't remember correctly) to 1k and later we can adjust this better and add visibility into the memory consumption.